design/kubernetes-friendly-oam.md

```diff
@@ -0,0 +1,478 @@
+    required: false
+    fieldPaths:
+    - ".spec.engineVersion"
+```
+
+The above WorkloadDefinition and Component leverages the existing Crossplane
+RedisCluster resource claim. Crossplane as a tentative OAM runtime need not
+define an extended workload type; it need only create a WorkloadDefinition
+pointing to [the existing CustomResourceDefinition] of the RedisCluster resource
+claim.
+
+The Component design enables OAM runtimes to break the task of reconciling the
+desired state of an application with the actual state of the world into two
+distinct stages:
+
+1. Render workloads. The OAM runtime iterates over the Components referenced by
+   an ApplicationConfiguration, instantiating a distinct workload resource for
+   each component. In the above examples this would involve creating a distinct
+   Server resource, and a distinct RedisCluster resource.
+2. Reconcile workloads. By producing valid instances of existing Kubernetes
+   custom resources OAM is able to leverage existing controllers. For example
+   the existing Crossplane RedisClaim controller(s) would be responsible for
+   reconciling the RedisCluster custom resource produced during the render
+   phase.
+
+Note that this two phase reconciliation and the production of distinct workload
+resources would not be required by the OAM specification. Runtimes could choose
+to iterate over the referenced components, render them in memory, and act upon
+the rendered components directly.
+
+### Traits
+
+Like workloads, each trait is of a distinct kind. A TraitDefinition defines the
+schema of each kind of trait, and limits what workload kinds it applies to:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha1
+kind: TraitDefinition
+metadata:
+  name: manualscalertrait.core.oam.dev
+spec:
+  appliesToWorkloads:
+    - server.core.oam.dev/v1alpha1
+    - worker.core.oam.dev/v1alpha1
+    - task.core.oam.dev/v1alpha1
+  definitionRef:
+    name: manualscalertrait.core.oam.dev
+```
+
+The above TraitDefinition references a Kubernetes CustomResourceDefinition named
+manualscalertrait.core.oam.dev, designating custom resources defined by said CRD
+to be valid OAM traits. As with WorkloadDefinitions, OAM runtimes that are not
+powered by Kubernetes could opt to either implement the CRD concept, or
+introduce a new field to the TraitDefinition spec that defines a kind of trait
+and its schema. Application operators can discover available OAM trait kinds by
+running `kubectl get traitdefinitions`. Traits would most frequently be defined
+by OAM runtimes or infrastructure operators.
+
+This document proposes that trait configuration be applied by be embedding one
+or more of a particular kind of trait under the `traits` array of the component
+it should apply to. For example:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha1
+kind: ApplicationConfiguration
+metadata:
+  name: cool-example
+spec:
+  components:
+  - componentName: example-server
+    traits:
+    - apiVersion: core.oam.dev/v1alpha1
```

Comment:
We should put a trait CR under an envelope field:

```
traits:
- name: manual-scaler
  instance:
    ...
```

In this way, we could achieve extensibility and add other models (ref/properties) later.

Reply:
I like the idea of using an envelope field, though I'd prefer to avoid duplicating the trait's name (we could already use its object metadata for that). I'll make an update.

Reply:
Wouldn't it make more sense if the trait was referring to a TraitDefinition? I'm imagining a workflow where an app operator asks the runtime for all available trait definitions, and then adds traits in the app config by picking from that list. In that case, they'd probably be looking at the name metadata field in the TraitDefinition. The runtime would also need to know the TraitDefinition to check if the trait supports the workload type of the component it's being applied to.

Reply:
@vturecek I think Nic's current example also make sense. e.g. if the TraitDefinition is referring to a CRD, then the Operator could embed a CR in this section (inside the envelope).

Reply:
> I like the idea of using an envelope field

I like this idea in principle, but it feels awkward in practice. See the below examples where I've tried embedding the CR under the key `trait` and `instance`.

```yaml
apiVersion: core.oam.dev/v1alpha1
kind: ApplicationConfiguration
metadata:
  name: cool-example
spec:
  components:
  - componentName: example-server
    traits:
      trait:
      - apiVersion: core.oam.dev/v1alpha1
        kind: ManualScalerTrait
        spec:
          replicaCount: 3
```

```yaml
apiVersion: core.oam.dev/v1alpha1
kind: ApplicationConfiguration
metadata:
  name: cool-example
spec:
  components:
  - componentName: example-server
    traits:
      instance:
      - apiVersion: core.oam.dev/v1alpha1
        kind: ManualScalerTrait
        spec:
          replicaCount: 3
```

> In this way, we could achieve extensibility and add other models (ref/properties) later.

I think this is really the only case in which it makes sense to put the trait CR under an envelope field. In any other case we have an extensibility story by adding fields to the trait CRs themselves. How confident are we that we'll want to go down this path some day? I still prefer to avoid it, because I think the spec will be simpler if there's only one way to do things (like use traits).

Reply:
> Wouldn't it make more sense if the trait was referring to a TraitDefinition?

It's not immediately obvious to me that it would make more sense - I think the UX of authoring ApplicationConfiguration is better when we're embedding a trait (not referencing a TraitDefinition), but I can see your point about trait discoverability. One thing we could do to help with this would be to use [custom printer columns](https://book.kubebuilder.io/reference/markers.html) for our `TraitDefinition` CRD such that folks running kubectl would see something like:


```
$ kubectl get traitdefinitions
NAME                  KIND               API-VERSION
manualscaler.oam.dev  ManualScalerTrait  core.oam.dev/v1alpha1
```

Reply:
Why not embed like this?

```yaml
apiVersion: core.oam.dev/v1alpha1
kind: ApplicationConfiguration
metadata:
  name: cool-example
spec:
  components:
  - componentName: example-server
    traits:
      - name: manual-scaler
        instance:
          apiVersion: core.oam.dev/v1alpha1
          kind: ManualScalerTrait
          spec:
            replicaCount: 3
```

Reply:
> Why not embed like this?

We discussed this in the OAM community meeting this morning, but for posterity:

* I think `name` is redundant, whether it's meant to be the name of the trait definition or the name of the trait instance that will be created. The former isn't needed at the ApplicationConfiguration level - the AppConfig controller will just instantiate the templated trait CR, whose separate controller will check its TraitDefinition. The latter duplicates the trait's existing `.metadata.name` field.
* I feel like the name `instance` for an envelope key "leaks implementation details" of the spec debate process, or put differently, doesn't make much sense to those of us who haven't worked on the spec.

We agreed that we do want to use an envelope field just in case, and that we'd name that field `trait`. I'll make the updat.e